### PR TITLE
NETOBSERV-117: show loading/empty/error state in netflow table

### PR DESCRIPTION
**Pull request by mariomac**
Shows empty and error states according to [PatternFly guidelines](https://www.patternfly.org/v4/components/table/react-demos/#empty-states).

This PR does not complete [NETOBSERV-117](https://issues.redhat.com/browse/NETOBSERV-117). When [NETOBSERV-72](https://github.com/netobserv/network-observability-console-plugin/pull/23) and [NETOBSERV-74](https://issues.redhat.com/browse/NETOBSERV-74) are completed, the "clear all filters" button will implement the supposed functionality.

image

image

It also adds a spinning icon that is shown when the page is loaded/refreshed completely but not when the table is refreshed through the dropdown:

https://user-images.githubusercontent.com/939550/147267936-4fb66656-6fbc-438e-88e9-ed5b7cc150da.mp4

### CODE REVIEW

**Review by OlivierCazade — APPROVED** [review]

File: web/src/components/netflow-traffic.tsx [hunk 1]
```diff
@@ -25,9 +25,9 @@ import './netflow-traffic.css';
 
 export const NetflowTraffic: React.FC = () => {
   const [extensions] = useResolvedExtensions<ModelFeatureFlag>(isModelFeatureFlag);
-  const [loading, setLoading] = React.useState(false);
+  const [loading, setLoading] = React.useState(true);
   const [flows, setFlows] = React.useState<ParsedStream[]>([]);
-  const [error, setError] = React.useState(undefined);
+  const [error, setError] = React.useState<string>(undefined);
```
Comment (jpinsonneau) [inline web/src/components/netflow-traffic.tsx:30]: this should be <string | undefined>

File: web/src/components/netflow-table.tsx [hunk 2]
```diff
@@ -94,12 +115,72 @@ const NetflowTable: React.FC<{
...
+  colSpan: number;
+}> = ({ colSpan }) => {
+  return (
+    <Tr>
+      <Td colSpan={colSpan}>
+        <Bullseye>
+          <EmptyState id="no-results-found" variant={EmptyStateVariant.small}>
+            <EmptyStateIcon icon={SearchIcon} />
+            <Title headingLevel="h2" size="lg">
+              No results found
```
Comment (jpinsonneau) [inline web/src/components/netflow-table.tsx:156]: you need to call translation here

Check `const { t } = useTranslation('plugin__network-observability-plugin');` and call t(`No results found`) after adding your text in `plugin__network-observability-plugin.json`

File: web/src/components/netflow-table.tsx [hunk 3]
```diff
@@ -94,12 +115,72 @@ const NetflowTable: React.FC<{
...
+            </Title>
+            <EmptyStateBody>Clear all filters and try again.</EmptyStateBody>
```
Comment (jpinsonneau) [inline web/src/components/netflow-table.tsx:158]: same for translation

File: web/src/components/netflow-table.tsx [hunk 4]
```diff
@@ -94,12 +115,72 @@ const NetflowTable: React.FC<{
...
+            {/*TODO: take onClick action after task NETOBSERV-74 is completed*/}
+            <Button variant="link">Clear all filters</Button>
```
Comment (jpinsonneau) [inline web/src/components/netflow-table.tsx:160]: same for translation

File: web/src/components/netflow-table.tsx [hunk 5]
```diff
@@ -94,12 +115,72 @@ const NetflowTable: React.FC<{
...
+const ErrorState: React.FC<{
+  colSpan: number;
+  error: string;
+}> = ({ colSpan, error }) => {
+  return (
+    <Tr>
+      <Td colSpan={colSpan}>
+        <EmptyState variant={EmptyStateVariant.small}>
+          <Title headingLevel="h2" size="lg">
+            Unable to get flows
```
Comment (jpinsonneau) [inline web/src/components/netflow-table.tsx:177]: same for translation

File: web/src/components/__tests__/netflow-table.spec.tsx [hunk 6]
```diff
@@ -49,4 +47,42 @@ describe('<NetflowTable />', () => {
...
+    wrapper.update();
+    expect(wrapper.find('LoadingContents')).toHaveLength(0);
+    expect(wrapper.find('NoResultsFound')).toHaveLength(0);
+    expect(wrapper.find('ErrorState')).toHaveLength(0);
+    expect(wrapper.find(NetflowTableRow)).toHaveLength(FlowsSample.length);
+  });
+  it('should render a spinning slide and then a NoResultsFound message if no flows are found', () => {
+    const wrapper = mount(<NetflowTable loading={true} flows={[]} setFlows={setFlows} columns={ColumnsSample} />);
+    expect(wrapper.find(NetflowTable)).toBeTruthy();
+    expect(wrapper.find('LoadingContents')).toHaveLength(1);
```
Comment (jpinsonneau) [inline web/src/components/__tests__/netflow-table.spec.tsx:67]: what is the difference here between calling find on a React.FC like `NetflowTableRow` and on a text like `'ErrorState'` ?

**Review by jpinsonneau — COMMENTED** [review]
As [discussed on slack](https://coreos.slack.com/archives/C02939DP5L5/p1640272331167000), it should be good to have the spinner showing when coming from no data to another table state

File: web/src/components/netflow-table.tsx (hunk 2 above)
Comment (mariomac) [reply web/src/components/netflow-table.tsx:156]: Ah yeah! I forgot it. Thanks

File: web/src/components/__tests__/netflow-table.spec.tsx (hunk 6 above)
Comment (mariomac) [reply web/src/components/__tests__/netflow-table.spec.tsx:67]: 'ErrorState' is not exported so it's not visible from the tests. Removing the quotes would not allow the tests to compile: `ReferenceError: ErrorState is not defined`

Comment (jpinsonneau) [reply web/src/components/__tests__/netflow-table.spec.tsx:67]: Ho that's true. You created multiple `React.FC` in the same file.

For private usage, I would prefer [jsx const](https://github.com/netobserv/network-observability-console-plugin/blob/main/web/src/components/columns-modal.tsx#L102) in the same Function Component.

File: web/src/components/netflow-table.tsx [hunk 7]
```diff
@@ -94,9 +151,14 @@ const NetflowTable: React.FC<{
...
-          <NetflowTableRow key={i} flow={f} columns={columns} />
-        ))}
+        {error
+          ? errorState
+          : _.isEmpty(flows)
+          ? // empty flows might mean that page is loading or no flows found
+            loading
+            ? loadingContents
+            : noResultsFound
+          : flows.map((f, i) => <NetflowTableRow key={i} flow={f} columns={columns} />)}
```
Comment (mariomac) [inline web/src/components/netflow-table.tsx:161]: Maybe the nested conditions are a bit cumbersome but the npm formatter didn't allow me to organize it in another way 😥

Comment (jpinsonneau) [reply web/src/components/netflow-table.tsx:161]: What about using a `let` that will be filled according to if / else conditions like : 
```
let bodyContent;
if(error) {
   bodyContent = (
    <Tr>
      <Td colSpan={columns.length}>
        <EmptyState data-test="error-state" variant={EmptyStateVariant.small}>
          <Title headingLevel="h2" size="lg">
            {t('Unable to get flows')}
          </Title>
          <EmptyStateBody>{error}</EmptyStateBody>
        </EmptyState>
      </Td>
    </Tr>
  );
} else if(_.isEmpty(flows)) {
   if(loading) {
      bodyContent = (
        <Tr>
          <Td colSpan={columns.length}>
            <Bullseye data-test="loading-contents">
              <Spinner size="xl" />
            </Bullseye>
          </Td>
        </Tr>
      );
   } else {
      bodyContent = (
        <Tr>
          <Td colSpan={columns.length}>
            <Bullseye data-test="no-results-found">
              <EmptyState variant={EmptyStateVariant.small}>
                <EmptyStateIcon icon={SearchIcon} />
                <Title headingLevel="h2" size="lg">
                  {t('No results found')}
                </Title>
                <EmptyStateBody>{t('Clear all filters and try again.')}</EmptyStateBody>
                {/*TODO: take onClick action after task NETOBSERV-74 is completed*/}
                <Button variant="link">{t('Clear all filters')}</Button>
              </EmptyState>
            </Bullseye>
          </Td>
        </Tr>
      );
   }
} else {
   bodyContent = flows.map((f, i) => <NetflowTableRow key={i} flow={f} columns={columns} />);
}
```

Then simply : 
`<Tbody>{bodyContent}</Tbody>` in the return statement

Comment (mariomac) [reply web/src/components/netflow-table.tsx:161]: Fine with me. I observed that, in the console code, if-elses were unused in favor of the ternary operator so I thought it was some kind on undesirable idiom.

**Comment by jpinsonneau** [comment]
LGTM, just need to rebase, and bind the `Clear all filters` button from `bodyContent`

You can move `clearAll` const from `filters-toolbar.tsx` to `netflow-traffic.tsx` and send it through props to share it between React.FC

**Review by jotak — APPROVED** [review]
/lgtm
(code lgtm, not tested)

**Pull request merged**